presto-spi/src/main/java/io/prestosql/spi/type/TimeZoneKey.java

```diff
@@ -210,102 +251,89 @@ public static boolean isUtcZoneId(String zoneId)
+                    minute = Integer.parseInt(lowerCased, colon + 1, lowerCased.length(), 10);
+                }
+                else if (lowerCased.length() - offset <= 3) {
+                    // +H or +HH
+                    hour = Integer.parseInt(lowerCased, offset, lowerCased.length(), 10);
+                }
+                else {
+                    // let the JDK handle it below
+                    break;
+                }
+
+                if (hour == 0 && minute == 0) {
+                    return "utc";
+                }
+
+                if (prefix.equals("etc/gmt")) {
+                    hour = -hour;
+                }
 
-        // In some zones systems, these will start with UTC, GMT or UT.
-        int length = zoneId.length();
-        boolean startsWithEtcGmt = false;
-        if (length > 3 && (zoneId.startsWith("utc") || zoneId.startsWith("gmt"))) {
-            if (startsWithEtc && zoneId.startsWith("gmt")) {
-                startsWithEtcGmt = true;
+                return formatZoneOffset(hour, minute);
             }
-            zoneId = zoneId.substring(3);
-            length = zoneId.length();
-        }
-        else if (length > 2 && zoneId.startsWith("ut")) {
-            zoneId = zoneId.substring(2);
-            length = zoneId.length();
         }
 
-        // (+/-)00:00 is UTC
-        if ("+00:00".equals(zoneId) || "-00:00".equals(zoneId)) {
-            return "utc";
+        // Normalize using JDK rules. In particular, normalize forms such as +0735, +7, +073521
+        try {
+            zoneId = ZoneId.of(zoneId).getId();
         }
-
-        // if zoneId matches XXX:XX, it is likely +HH:mm, so just return it
-        // since only offset time zones will contain a `:` character
-        if (length == 6 && zoneId.charAt(3) == ':') {
+        catch (Exception e) {
             return zoneId;
         }
 
-        //
-        // Rewrite (+/-)H[H] to (+/-)HH:00
-        //
-        if (length != 2 && length != 3) {
-            return originalZoneId;
+        if (isUtcEquivalentName(zoneId)) {
+            return "utc";
         }
 
-        // zone must start with a plus or minus sign
-        char signChar = zoneId.charAt(0);
-        if (signChar != '+' && signChar != '-') {
-            return originalZoneId;
-        }
-        if (startsWithEtcGmt) {
-            // Flip sign for Etc/GMT(+/-)H[H]
-            signChar = signChar == '-' ? '+' : '-';
-        }
+        return zoneId;
+    }
 
-        // extract the tens and ones characters for the hour
-        char hourTens;
-        char hourOnes;
-        if (length == 2) {
-            hourTens = '0';
-            hourOnes = zoneId.charAt(1);
-        }
-        else {
-            hourTens = zoneId.charAt(1);
-            hourOnes = zoneId.charAt(2);
-        }
+    private static String formatZoneOffset(int hour, int minute)
+    {
+        StringBuilder builder = new StringBuilder();
```

Comment:
6 capacity, if you care about performance

(`format` otherwise)

Reply:
`format` is much slower. Default StringBuilder capacity is 16, so performance-wise, setting it to 6 doesn't matter.